### PR TITLE
We eat basic backend

### PR DESCRIPTION
**Pull request by dotanlang**
Restaurants controller and model with validations and tests

### CODE REVIEW

File: app/models/restaurant.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+
+class NameValidator < ActiveModel::Validator
```
Comment (ooffeerr) [inline app/models/restaurant.rb:2]: Not a mandatory thing, but we usually respect each class with it's own file.

File: db/seeds.rb [hunk 2]
```diff
@@ -5,3 +5,7 @@
 #
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
+Restaurant.delete_all
+Restaurant.create(name: 'Shi-Shi', cuisine: 'asian', rating: 3, ten_bis: false, address: 'Shlomo Ibn Gabirol St 33, Tel Aviv-Yafo', max_delivery_time: 40)
```
Comment (ooffeerr) [inline db/seeds.rb:9]: 👍

File: spec/controllers/restaurants_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+  describe '#create' do
+    it "creates a new restaurant" do
+      post :create, params: { name: 'Artzieli11',
+                                    cuisine: 'pizza',
+                                    rating: 3,
+                                    ten_bis: true,
+                                    address: 'street in tel aviv',
+                                    max_delivery_time: 45 }
+      expect(response.status).to eq(200)
```
Comment (ooffeerr) [inline spec/controllers/restaurants_controller_spec.rb:21]: You say 'create' creates a new restaurant, but you only validate the http status - not the creation of the new restaurant.
From what I see this is something you're doing through the Controller's spec - an HTTP result does not always mean an actual change to some DB entry.

File: spec/controllers/restaurants_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                              rating: 3,
+                              ten_bis: true,
+                              address: 'street in tel aviv',
+                              max_delivery_time: 45 }
+      expect(response.status).to eq(409)
+    end
+
+  end
+
+  describe '#update' do
```
Comment (ooffeerr) [inline spec/controllers/restaurants_controller_spec.rb:64]: A very comprehensive test suite! Well done!

File: spec/controllers/restaurants_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    it "fails to update a non existent id" do
+      expect {
+        patch :update, params: { id: 22, name: 'Chez Pierre' }
+      }.to raise_error(ActiveRecord::RecordNotFound)
+    end
+  end
+
+  describe '#index' do
+    let!(:rest1) { FactoryBot.create(:restaurant) }
```
Comment (ooffeerr) [inline spec/controllers/restaurants_controller_spec.rb:81]: We usually try not to abuse the bang version of let, as it enforces creation at a non lazy way. We are all lazy here.

File: spec/controllers/restaurants_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      expect(response.status).to eq(200)
+    end
+
+    it "can't create a restaurant without a name" do
+      post :create, params: { address: 'street in tel aviv' }
+      expect(response.status).to eq(400)
+    end
+
+    it "can't create a restaurant with a name that already exists" do
+      post :create, params: { name: 'Artzieli1',
```
Comment (ooffeerr) [inline spec/controllers/restaurants_controller_spec.rb:45]: Another way could be letting FactoryBot doing all params work for you. Something like 
```
let (:rest) {FactoryBot.build(:restaurant)} # notice I used *build* which does not persist this instance
      post :create, params: { rest.to_params } # not sure this is the syntax for object->params, but you get the idea...
```

File: spec/factories/restaurants.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+require 'factory_bot'
+
+FactoryBot.define do
```
Comment (ooffeerr) [inline spec/factories/restaurants.rb:3]: YES!
Great builder! Hope it saved you the gritty at the tests. Did you use all the traits (I couldn't see them at the test)? Where they helpful?

File: spec/controllers/restaurants_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    end
+
+    it "can't create a restaurant with an invalid rating" do
+      post :create, params: { name: 'Artzieli',
+                              cuisine: 'pizza',
+                              rating: 7,
+                              ten_bis: true,
+                              address: 'street in tel aviv',
+                              max_delivery_time: 45 }
+      expect(response.status).to eq(400)
```
Comment (ooffeerr) [inline spec/controllers/restaurants_controller_spec.rb:31]: Isn't there a constant for these HTTP values?

**Review by ooffeerr — APPROVED** [review]
My name is Ofer Schonberger and I approve this PR. But with some comments.

File: app/models/restaurant.rb (hunk 1 above)
Comment (dotanlang) [reply app/models/restaurant.rb:2]: done. moved to a different file called restaurant_name_validation

**Pull request merged**